docs/customtut_parsers.rst

```diff
@@ -457,125 +457,202 @@ The following shows our completed parser including documentation.
-       import doctest
-       doctest.testmod()
+        def last(self, keyword):
+            """str: Returns the value of the last keyword found in config."""
+            entries = self.__getitem__(keyword)
+            if entries:
+                return entries[-1]
 
 Parser Testing
 --------------
 
-It is important that we ensure our tests will run successfully after any change
-to our parser. We are able to do that in two ways, first by running ``doctest``
-to test our *Examples* section of the ``ssh`` module, and second by running
-``pytest``.
+It is important that we ensure our tests will run successfully after any
+change to our parser. We are able to do that in two ways, first by using
+``doctest`` to test our *Examples* section of the ``ssh`` module, and second
+by writing tests that can be tested automatically using ``pytest``.  Starting
+with adding ``import doctest`` our original code:
 
-``doctest`` is implemented by including the following lines in our module::
+.. code-block:: python
+    :linenos:
+
+    from insights.parsers import ssh
+    from insights.tests import context_wrap
+    import doctest
 
-   if __name__ == '__main__':
-       import doctest
-       doctest.testmod()
+    SSHD_CONFIG_INPUT = """
+    #	$OpenBSD: sshd_config,v 1.93 2014/01/10 05:59:19 djm Exp $
 
-To execute the ``doctest`` use the following command::
+    Port 22
+    #AddressFamily any
+    ListenAddress 10.110.0.1
+    Port 22
+    ListenAddress 10.110.1.1
+    #ListenAddress ::
 
-    $ python -m insights.parsers.ssh
+    # The default requires explicit activation of protocol 1
+    #Protocol 2
+    Protocol 1
+    """
 
-If no errors are displayed then ``doctest`` was successful. To run
-``pytest`` on just the ``ssh`` parser execute the following command::
+    def test_sshd_config():
+        sshd_config = SshDConfig(context_wrap(SSHD_CONFIG_INPUT))
+        assert sshd_config is not None
+        assert 'Port' in sshd_config
+        assert 'PORT' in sshd_config
+        assert sshd_config['port'] == ['22', '22']
+        assert 'ListenAddress' in sshd_config
+        assert sshd_config['ListenAddress'] == ['10.110.0.1', '10.110.0.1']
+        assert sshd_config['Protocol'] == ['1']
+        assert 'AddressFamily' not in sshd_config
+        ports = [l for l in sshd_config if l.keyword == 'Port']
+        assert len(ports) == 2
+        assert ports[0].value == '22'
+
+To test the documentation, we can then use ``doctest``:
+
+.. code-block:: python
+    :linenos:
+    :lineno-start: 35
+
+    def test_sshd_documentation():
+        """
+        Here we test the examples in the documentation automatically using
+        doctest.  We set up an environment which is similar to what a
+        rule writer might see - a 'shared' object that stores parser
+        objects by class.  This saves doing this setup in the example code.
+        """
+        env = {
+            'SshDConfig': ssh.SshDConfig,
+            'shared': {
```

Comment:
3.x no longer uses the `shared` mechanism so we should change this to reflect how it should be done in 3.x.